Support CPU profiling sections of code

breeswish opened the pull request:
Signed-off-by: Breezewish <breezewish@pingcap.com>

<!--
Thank you for contributing to TiKV! Please read TiKV's [CONTRIBUTING](https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

## What have you changed? (mandatory)

We usually need to find out why specific sections of code are slow. However, sometimes it cannot be done via simple perf, i.e. when there is a bootstrap. This PR provides facility to programmatically start or stop CPU profiling. In this way, we can start profiling before our interested code begins and stop profiling after it ends.

We can explore more interesting usages in the future, i.e. toggle profiling dynamically via signals, or environment variables. Future PRs are welcome!

The manual is [in the code](https://github.com/tikv/tikv/blob/9ddf5e91a7d8ee1a7857c86ac828afb00716d931/components/profiler/src/lib.rs). I paste it here for easy reading:

---

Profile a part of the code using CPU Profiler from gperftools or Callgrind.

## Requirements

1. Linux

   Other OS may also work however, not tested.

2. gperftools

   You can follow its [INSTALL manual](https://github.com/gperftools/gperftools/blob/master/INSTALL).
   Roughly the instructions are the following:

   1. Download packages from [release](https://github.com/gperftools/gperftools/releases)
   2. Run `./configure`
   3. Run `make install`

## Usage

```rust
profiler::start("./app.profile");
some_complex_code();
profiler::stop();
```

Then, compile the code with `profiling` feature enabled and run the code with environment
variable `TIKV_PROFILE=1`.

By default, a profile called `app.profile` will be generated by CPU Profiler.
You can then analyze the profile using [pprof](https://github.com/google/pprof).

If the application is running in Callgrind, a Callgrind profile dump will be generated instead.
Notice that you should run Callgrind with command line option `--instr-atstart=no`, e.g.:

```bash
TIKV_PROFILE=1 valgrind --tool=callgrind --instr-atstart=no ./my_example
```

Also see `examples/prime.rs`.

breeswish commented:
/rebuild

zhouqiang-cl commented:
/rebuild

siddontang commented:
@breeswish I think you can trigger it through HTTP API now.

siddontang commented:
Do we need to install gperftools at first?

breeswish commented:
@siddontang Yes. I copied the manual from code to the PR description now. I also updated the code to support Callgrind (so that we can profile cache hit etc).

zhouqiang-cl commented:
/rebuild

zhouqiang-cl commented:
/rebuild

siddontang commented:
So if the user doesn't install perftools, can it build TiKV now? 

I suggest you introduce valgrind in another PR.

breeswish commented:
@siddontang Yes. Only when profiling feature is enabled, gperftools is required.

The support of Valgrind is only 4 LOC, I think it's not a big problem.

siddontang commented:
do we need to install vagrind manally too?

siddontang commented:
IMO, if we build with this feature, we don't need using environment forcibly to control it. Only calling start and stop is enough.

zhouqiang-cl commented:
/rebuild

breeswish commented:
Updated. No need to set "TIKV_PROFILE=1" now.

breeswish commented:
@brson PTAL for this~ Thanks a lot!

brson commented:
This is very cool @breeswish.

I guess that the idea is to insert start and stop temporarily and then remove them when done? Otherwise there's a lot of deadlock potential.

One nice change would be to change the profiler manifest to make all the dependencies optional (particularly the profiling deps), and have the `profiling` target enable them. That would save the download, build, and link of the profiler crates.

Is there anything holding up landing this?

Maybe docs - like the allocator configuration this seems like something that should be surfaced in a developer's guide. Not necessary for this PR though.

kennytm commented on components/profiler/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "profiler"
+version = "0.0.1"
+publish = false
```
This new package should use `edition = 2018`?

breeswish commented:
@brson @kennytm Thanks a lot for reviewing! I have updated my `Cargo.toml` to make these dependencies optional (not sure it there are better ways). Also updated docs for MacOS instructions.

> I guess that the idea is to insert start and stop temporarily and then remove them when done? Otherwise there's a lot of deadlock potential.

Yes. A better way maybe, provide interfaces to trigger a start and stop it moment later so that a profile can be generated. This does not introduce runtime costs when profiling is not started.

> Is there anything holding up landing this?

Nop, it's just lack of reviewers previously :)

siddontang commented:
@brson 

I think we can advance this PR now. 
But here I prefer using HTTP API to control profile.

breeswish commented:
@siddontang We can have another PR that utilize the interface via HTTP API.

siddontang commented:
PTAL @brson

brson commented:
@siddontang acknowledged that we want an HTTP interface, and that this is a good start.

brson commented:
lgtm

BusyJay commented on components/profiler/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+publish = false
+
+[features]
+profiling = ["lazy_static", "cpuprofiler", "callgrind", "valgrind_request"]
```
Can it compile on Windows without the feature enabled?

breeswish replied:
It should work though I don't have a Windows machine. I changed it to `[target.'cfg(linux)'.dependencies]` and it can compile on my MacOS.

BusyJay commented on components/profiler/src/lib.rs:
```diff
@@ -0,0 +1,58 @@
+//! Also see `examples/prime.rs`.
+
+#[allow(unused_extern_crates)]
+extern crate tikv_alloc;
```
Why use `tikv_alloc`?

brson replied:
tikv_alloc generally needs to be linked into every crate that doesn't link to tikv, so that tests and benches of that crate use tikv's allocator. I don't _think_ this extern crate statement is needed though as long as the dependency exists.

breeswish replied:
CI will fail if alloc is not linked

brson replied:
Ah, right. There are tests that all binaries contain jemalloc.

BusyJay commented on components/profiler/src/lib.rs:
```diff
@@ -0,0 +1,58 @@
+extern crate tikv_alloc;
+
+#[cfg(all(unix, feature = "profiling"))]
+mod profiler_linux;
```
Should it be named as `profiler_unix`?

BusyJay commented on components/profiler/src/profiler_linux.rs:
```diff
@@ -0,0 +1,68 @@
+
+    // Profiling in progress.
+    if *profiler != Profiler::None {
+        return;
```
Should it return error instead? For example, what if two profiling requests are received at the same time?

BusyJay commented on components/profiler/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+[target.'cfg(unix)'.dependencies]
+lazy_static = { version = "1.2.0", optional = true }
+cpuprofiler = { version = "0.0.3", optional = true }
+callgrind = { version = "1.1.0", optional = true }
```
Is valgrind a useful use case? I think most of the time only cpuprofiling is used.

breeswish replied:
valgrind is very useful for micro benchmark, which can report precise amount of function calls, as well as precise (emulated) cache hit.

brson replied:
I'm also hoping that this crate can be extended into a general purpose profiling module that can be published for the community, and ultimately fulfill the various requirements of the Go profiling tools. More profiler options in that case seems good.
